In [1]:
from rdkit.Chem import AllChem as Chem
from rdkit.Chem import Draw
from rdkit.Chem.Draw import IPythonConsole

In [2]:
import csv
with open('Tomas_Otava_CDG_Epidermal_growth_factor_receptor_erbB1_list_of_compounds.csv','r') as csvfile:
    reader = csv.DictReader(csvfile,delimiter=";")
    rawdata = [x for x in reader]
lig_MOLECULES = [Chem.MolFromSmiles(x['Smiles']) for x in rawdata]

Načtení erbB1

In [3]:
suppl = Chem.SDMolSupplier('drugbank.sdf')
drugs = [m for m in suppl if m]
len(drugs)

RDKit ERROR: [18:46:11] Explicit valence for atom # 7 N, 4, is greater than permitted
RDKit ERROR: [18:46:11] ERROR: Could not sanitize molecule ending on line 20009
RDKit WARNING: [18:46:11] Warning: conflicting stereochemistry at atom 5 ignored.
RDKit WARNING: [18:46:11] Warning: conflicting stereochemistry at atom 5 ignored.
RDKit ERROR: [18:46:11] ERROR: Explicit valence for atom # 7 N, 4, is greater than permitted
RDKit ERROR: [18:46:11] Can't kekulize mol.  Unkekulized atoms: 1 2 3 5 6 7 8 9 10
RDKit ERROR: 
RDKit ERROR: [18:46:11] ERROR: Could not sanitize molecule ending on line 250947
RDKit ERROR: [18:46:11] ERROR: Can't kekulize mol.  Unkekulized atoms: 1 2 3 5 6 7 8 9 10
RDKit ERROR: 
RDKit ERROR: [18:46:11] Explicit valence for atom # 17 O, 3, is greater than permitted
RDKit ERROR: [18:46:11] ERROR: Could not sanitize molecule ending on line 258130
RDKit ERROR: [18:46:11] ERROR: Explicit valence for atom # 17 O, 3, is greater than permitted
RDKit ERROR: [18:46:11] Can't kek

RDKit ERROR: [18:46:12] ERROR: Can't kekulize mol.  Unkekulized atoms: 0 1 2 3 4 5 6 7 8
RDKit ERROR: 
RDKit ERROR: [18:46:12] Can't kekulize mol.  Unkekulized atoms: 3 4 5 6 8 9 10
RDKit ERROR: 
RDKit ERROR: [18:46:12] ERROR: Could not sanitize molecule ending on line 442989
RDKit ERROR: [18:46:12] ERROR: Can't kekulize mol.  Unkekulized atoms: 3 4 5 6 8 9 10
RDKit ERROR: 
RDKit ERROR: [18:46:12] Can't kekulize mol.  Unkekulized atoms: 3 4 5 6 9
RDKit ERROR: 
RDKit ERROR: [18:46:12] ERROR: Could not sanitize molecule ending on line 447465
RDKit ERROR: [18:46:12] ERROR: Can't kekulize mol.  Unkekulized atoms: 3 4 5 6 9
RDKit ERROR: 
RDKit ERROR: [18:46:12] Can't kekulize mol.  Unkekulized atoms: 16 17 18 19 20
RDKit ERROR: 
RDKit ERROR: [18:46:12] ERROR: Could not sanitize molecule ending on line 466177
RDKit ERROR: [18:46:12] ERROR: Can't kekulize mol.  Unkekulized atoms: 16 17 18 19 20
RDKit ERROR: 
RDKit ERROR: [18:46:12] Can't kekulize mol.  Unkekulized atoms: 7 8 9 10 17
RDKit ERR

RDKit ERROR: [18:46:12] ERROR: Can't kekulize mol.  Unkekulized atoms: 0 1 2 4 5 6 7 8 9 10
RDKit ERROR: 
RDKit ERROR: [18:46:13] Explicit valence for atom # 0 Be, 3, is greater than permitted
RDKit ERROR: [18:46:13] ERROR: Could not sanitize molecule ending on line 995692
RDKit ERROR: [18:46:13] ERROR: Explicit valence for atom # 0 Be, 3, is greater than permitted
RDKit ERROR: [18:46:13] Explicit valence for atom # 0 C greater than permitted
RDKit ERROR: [18:46:13] ERROR: Could not sanitize molecule ending on line 1151592


7113

Načtení drugbanku

In [4]:
fp3344 = Chem.GetMorganFingerprintAsBitVect(lig_MOLECULES[3344], 2, nBits=1024)

In [5]:
fp3344.ToBitString()

'000000000000000000000000001000000100000100000010000000000000000000000000010000001000000000000000000000000000000000000001010000001000000000000000000000000000000000000000000000000000000000000001001000000010000000000000100000000000000000000100000000010000000000000000000000000000000000000000000000100000011000000000100000000010000001000000000000010000000000001000000000100000000000100000000000000000000000000000000000000000000000001000000000000000000000000000101000001100000000000000000000000101000000000000000000001000000000001000000000000000000000000000000001000100000000000000000000000000000000000000100000000000000000000000000000000001000000000000000000000001000000000000000000000000000000000001000100000000100000000000000000100000000000000100000000000000000000000001000000000000010000100000001000000000000000000000100010000000000000000000010100000100000000000000000000000001000000000000100000000000000010000000000000000000001000000001000000000010000000000001100000000000000000000000000000000000000

Opět moje oblíbená zcela náhodná molekula 3344

In [6]:
erbB1_ligands_ECFP = [Chem.GetMorganFingerprintAsBitVect(x, 2, nBits=1024)
                     for x in lig_MOLECULES]
drugs_ECFP = [Chem.GetMorganFingerprintAsBitVect(x, 2, nBits=1024) for x in drugs]

In [7]:
erbB1_ligands_ECFP_sums = [0]*erbB1_ligands_ECFP[0].GetNumBits()
for key in erbB1_ligands_ECFP:
    for onbit in key.GetOnBits():
        erbB1_ligands_ECFP_sums[onbit] += 1
erbB1_ligands_ECFP_sums

[50,
 1403,
 327,
 610,
 1368,
 100,
 60,
 316,
 1022,
 218,
 644,
 315,
 61,
 1498,
 1213,
 4729,
 96,
 25,
 83,
 279,
 196,
 323,
 55,
 196,
 78,
 753,
 359,
 187,
 73,
 467,
 36,
 1036,
 695,
 8911,
 959,
 45,
 1177,
 72,
 75,
 323,
 135,
 801,
 1428,
 424,
 48,
 907,
 431,
 59,
 81,
 62,
 176,
 229,
 54,
 529,
 144,
 400,
 341,
 197,
 457,
 181,
 180,
 805,
 296,
 721,
 7875,
 878,
 61,
 73,
 109,
 107,
 51,
 153,
 342,
 5646,
 711,
 839,
 203,
 1027,
 171,
 259,
 7275,
 96,
 141,
 295,
 825,
 294,
 858,
 331,
 71,
 150,
 2478,
 34,
 228,
 22,
 1182,
 78,
 496,
 287,
 185,
 73,
 91,
 71,
 712,
 56,
 361,
 91,
 201,
 926,
 40,
 141,
 293,
 111,
 934,
 450,
 1168,
 362,
 611,
 534,
 270,
 1194,
 55,
 3051,
 201,
 176,
 60,
 238,
 202,
 91,
 9759,
 323,
 386,
 431,
 142,
 3659,
 244,
 239,
 2312,
 22,
 453,
 1454,
 768,
 67,
 201,
 160,
 289,
 391,
 119,
 2435,
 62,
 349,
 106,
 219,
 230,
 32,
 191,
 379,
 24,
 2714,
 441,
 83,
 224,
 1370,
 667,
 67,
 82,
 568,
 188,
 277,
 590,
 14

In [8]:
drugs_ECFP_sums = [0]*drugs_ECFP[0].GetNumBits()
for key in drugs_ECFP:
    for onbit in key.GetOnBits():
        drugs_ECFP_sums[onbit] += 1
drugs_ECFP_sums

[120,
 2360,
 285,
 317,
 939,
 231,
 29,
 138,
 213,
 303,
 260,
 258,
 68,
 665,
 227,
 828,
 103,
 104,
 93,
 187,
 118,
 203,
 63,
 237,
 153,
 162,
 127,
 111,
 62,
 211,
 69,
 172,
 158,
 4264,
 155,
 111,
 1076,
 73,
 87,
 207,
 79,
 440,
 648,
 112,
 78,
 261,
 319,
 76,
 52,
 97,
 89,
 123,
 245,
 271,
 100,
 63,
 105,
 85,
 158,
 178,
 69,
 179,
 204,
 219,
 2683,
 147,
 55,
 173,
 97,
 226,
 76,
 345,
 179,
 388,
 255,
 465,
 76,
 283,
 134,
 479,
 4939,
 169,
 152,
 122,
 313,
 87,
 270,
 124,
 107,
 105,
 1541,
 90,
 160,
 54,
 308,
 97,
 114,
 158,
 96,
 87,
 61,
 119,
 402,
 60,
 194,
 160,
 246,
 79,
 173,
 84,
 162,
 73,
 534,
 56,
 797,
 155,
 395,
 608,
 349,
 767,
 78,
 767,
 118,
 165,
 82,
 193,
 109,
 47,
 2503,
 89,
 337,
 123,
 98,
 162,
 127,
 52,
 625,
 56,
 515,
 399,
 667,
 59,
 125,
 150,
 289,
 198,
 99,
 2457,
 87,
 79,
 106,
 104,
 370,
 64,
 67,
 126,
 66,
 183,
 282,
 48,
 160,
 45,
 232,
 135,
 134,
 262,
 107,
 98,
 73,
 96,
 129,
 261,
 137,
 202,


In [9]:
erbB1_ligands_ECFP_scaled = [x/len(erbB1_ligands_ECFP) for x in erbB1_ligands_ECFP_sums]
drugs_ECFP_scaled = [x/len(drugs_ECFP) for x in drugs_ECFP_sums]
drugs_ECFP_scaled

[0.01687051876845213,
 0.3317868691128919,
 0.04006748207507381,
 0.044566287079994375,
 0.13201180936313792,
 0.03247574862927035,
 0.004077042035709265,
 0.01940109658371995,
 0.029945170814002532,
 0.04259805989034163,
 0.036552790664979615,
 0.036271615352172076,
 0.009559960635456208,
 0.09349079150850555,
 0.03191339800365528,
 0.1164065795023197,
 0.014480528609588077,
 0.014621116265991847,
 0.013074652045550401,
 0.026289891747504568,
 0.016589343455644594,
 0.028539294249964854,
 0.008857022353437369,
 0.03331927456769296,
 0.021509911429776464,
 0.022775200337410376,
 0.017854632363278503,
 0.01560522986081822,
 0.008716434697033601,
 0.029663995501194994,
 0.009700548291859975,
 0.024181076901448054,
 0.022212849711795303,
 0.5994657669056657,
 0.021791086742584002,
 0.01560522986081822,
 0.1512723182904541,
 0.010262898917475045,
 0.012231126107127795,
 0.029101644875579924,
 0.011106424855897652,
 0.06185856881765781,
 0.0911008013496415,
 0.015745817517221986,
 0.0109658

In [10]:
erbB1_drugbank_differences = [(i, a_b[0] - a_b[1])
                             for i, a_b in enumerate(zip(erbB1_ligands_ECFP_scaled, drugs_ECFP_scaled))]
erbB1_drugbank_differences.sort(key=lambda x: x[1])
erbB1_drugbank_differences[:3], erbB1_drugbank_differences[-3:]

([(1, -0.21951094081583683),
  (389, -0.16529313561416556),
  (147, -0.15056151583015212)],
 [(875, 0.511936382555394),
  (428, 0.5327988212238038),
  (491, 0.5673507691728891)])

Tak nakonec jsem dostal k těmto bitům kde pro DRUGBANK je favorizovaný bit 1 a pro erbB1 set 491.

Tak pojďme zjistit substruktury.

In [11]:
def substructures_of_bit_ecfp4_1024(mol, bit):
    bitinfo = {}
    Chem.GetMorganFingerprintAsBitVect(mol, 2, nBits=1024, bitInfo=bitinfo)
    submols = []
    for atom, radius in bitinfo.get(bit, []):
        if radius ==0:
            continue
        hit_atoms = Chem.FindAtomEnvironmentOfRadiusN(mol, radius, atom)
        submols.append(Chem.PathToSubmol(mol, hit_atoms))
    return submols

In [12]:
erbB1_substructure_smiles = {}
for x in lig_MOLECULES:
    for submol in substructures_of_bit_ecfp4_1024(x, 491):
        subsmiles = Chem.MolToSmiles(submol)
        erbB1_substructure_smiles[subsmiles] = erbB1_substructure_smiles.get(subsmiles, 0) + 1
erbB1_substructure_smiles

{'cNc': 7943,
 'cn(c)C(Cc)CO': 5,
 'ccc(cc)C(C)(C)N': 6,
 'cNC': 30,
 'CNc': 27,
 'cc([nH])N': 3,
 'C/C=C/CC': 7,
 'cc(-c)oc(C)n': 1,
 'cnc(N)c(c)C': 1,
 'CC(N)CN(c)c': 1,
 'CC(C)c': 1,
 'ccc(CO)c(c)O': 1,
 'CNCCN': 2,
 'C=C(C)C(=NO)C(=C)C': 6,
 'cnc([nH]c)SC': 2,
 'cn(-c)c(=O)c(n)=O': 1}

In [13]:
drugbank_substructure_smiles = {}
for x in drugs:
    for submol in substructures_of_bit_ecfp4_1024(x, 491):
        subsmiles = Chem.MolToSmiles(submol)
        drugbank_substructure_smiles[subsmiles] = drugbank_substructure_smiles.get(subsmiles, 0) + 1
drugbank_substructure_smiles

{'C=C/C=C/C': 9,
 'cC[C@H](NC)C(N)=O': 7,
 'cNc': 213,
 'NC(=O)CC(N)=O': 1,
 'CC(N)CN(c)c': 6,
 'C[NH3+]': 1,
 'CNc': 6,
 'CC(=O)N(C)O': 4,
 'cc(-c)oc(C)n': 1,
 'CC[C@@H](C)C(C)(C)C': 1,
 'cNC': 4,
 'C=c(n)ccc': 2,
 'cc(C)c(CO)c(C)n': 1,
 'CC(O)C(=O)O': 2,
 'C[C@H](O)CO': 1,
 'cC[C@@H](NC)C(N)=O': 2,
 'cc(n)[n+](C)cn': 1,
 'C[C@@H](N)C=O': 1,
 'cc(C)sc(n)n': 1,
 'ccc(c(c)C)C(c)[NH3+]': 2,
 'ccc(cc)[C@H](C)N': 1,
 'cc(c)C(CC)N(C)c': 1,
 'ccc(cc)[C@@H](C)N': 1,
 'CC(C)N(C)C(C)C': 1,
 'CC(C)N': 3,
 'ccc(Br)c(c)C': 2,
 'C[C@@H](N)CCS': 1,
 'cnn(Cc)c(c)=O': 1,
 'C/C=C/CC': 1,
 'cC(=N)CC(c)(C)N': 1,
 'C=C[C@@H](NC)[C@@H](C)O': 1,
 'CC[C@](C)(C(N)=O)C(C)C': 1,
 'cC(C)C': 1,
 'C[C@H](N)C[C@@H](C)O': 1,
 'ccc(c(c)C)c(c)c': 1,
 'NCCON': 1,
 'cc(c)NCC': 1}

Pro bit 491 je v setu erbB1 nejčastější substruktura cNc (7943), v drugbanku sice s menším zastoupením (213) je nejčastější tatáž substruktura.

In [14]:
erbB1_substructure_smiles = {}
for x in lig_MOLECULES:
    for submol in substructures_of_bit_ecfp4_1024(x, 1):
        subsmiles = Chem.MolToSmiles(submol)
        erbB1_substructure_smiles[subsmiles] = erbB1_substructure_smiles.get(subsmiles, 0) + 1
erbB1_substructure_smiles

{'CCCC#N': 4,
 'cc(C)[nH]c(c)c': 1,
 'ccc(C(c)=O)c(c)C': 20,
 'CN(C)C': 10,
 'cc(c)S': 62,
 'cccc(-c)n': 8,
 'C=C(C)C(=NC)NC': 12,
 'cc(s)NC(c)=O': 2,
 'ccc(C(c)=O)c(c)-c': 1,
 'CC[C@@H](O)[C@@H](C)O': 1,
 'cncc(n)N': 1,
 'CC(C)CN(C)C': 1,
 'cc(C)c(SC)n(c)C': 1,
 'CN(C)C(=O)CN': 1,
 'cncc(-c)s': 1,
 'cc(c)S(=O)(=O)NC': 1,
 'cnc(c(c)c)n(c)C': 2,
 'cc(c)N(C(c)=O)C(C)N': 1,
 'cc(c)-n(c(c)=O)c(c)c': 1,
 'cN(C)CC(N)F': 1}

In [15]:
drugbank_substructure_smiles = {}
for x in drugs:
    for submol in substructures_of_bit_ecfp4_1024(x, 1):
        subsmiles = Chem.MolToSmiles(submol)
        drugbank_substructure_smiles[subsmiles] = drugbank_substructure_smiles.get(subsmiles, 0) + 1
drugbank_substructure_smiles

{'ccc(C(C)=O)c(c)C': 11,
 'CN(C)C': 18,
 'cc(c)N(C(c)=O)C(C)N': 1,
 'CC(C)CN(C)C': 1,
 'CC[C@@H](O)[C@@H](C)O': 1,
 'cc(c)S(=O)(=O)NC': 2,
 'cc(=O)nc[nH]': 3,
 'cc(C)[nH]c(c)c': 2,
 'cC(=O)C[C@H](c)O': 3,
 'ccc(C(c)=O)c(c)C': 7,
 'cccc(-c)n': 11,
 'cCC(=O)CC': 2,
 'ccc(C(c)=O)c(c)-c': 2,
 'C=C[C@@H]([NH3+])C(=O)[O-]': 1,
 'CC(C)(C)C(=O)C(C)(C)C': 1,
 'cc(c)S': 24,
 'cc(c)CP(=O)(O)O': 1,
 'CC[C@H](O)[C@H](C)O': 1,
 'ccc[n+](c)C': 4,
 'CN(C)C(=O)CN': 2,
 'cc(c)-c(cn)cn': 1,
 'cC(=O)C[C@@H](c)O': 3,
 'ccc(c(c)-c)c(c)[oH+]': 2,
 'cnc(cn)[C@@H](C)S': 1,
 'CNC[P@H](=O)O': 1,
 'cccc(C)n': 1,
 'CC[C@H](O)C(=O)O': 1,
 'CN[C@@H](C(O)O)[C@@H](C)O': 1,
 'ccc(OC)c(c)C': 3,
 'cc(c)C=CC': 1,
 'CC[C@H](O)[C@@H](C)O': 1,
 'CC(C)(C)CC(C)(C)C': 1,
 'ccc(c(c)-c)c(c)[nH]': 2,
 'cc(s)NC(c)=O': 4,
 'cc(C)c(C)c(c)C': 1,
 'cccc(c)S': 1,
 'cncc(-c)s': 1,
 'CCCP(C)(C)=O': 1,
 'CCC(CC)N(C)c': 1,
 'cnc(c(-c)n)c(c)c': 1,
 'C=C(C)c(cc)c(c)C': 1}

Pro bit 1 je v setu erbB1 nejčastější substruktura cc(c)S (62), v drugbanku (24) je nejčastější opět tatáž substruktura.